### PR TITLE
Add ssdeep hash

### PR DESCRIPTION
**Pull request by wargio**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Adds ssdeep (aka. context triggered piecewise hash)

### CODE REVIEW

File: librz/hash/hash.c [hunk 1]
```diff
@@ -24,12 +25,45 @@ typedef struct hash_cfg_config_t {
 
 const static RzHashPlugin *hash_static_plugins[] = { RZ_HASH_STATIC_PLUGINS };
 
-RZ_API ut32 rz_hash_xxhash(RZ_NONNULL RzHash *rh, RZ_NONNULL const ut8 *input, size_t size) {
+RZ_API char *rz_hash_ssdeep(RZ_NONNULL const ut8 *input, size_t size) {
```
Comment (XVilka) [inline librz/hash/hash.c]: Doxygen and ownership, please

File: librz/hash/hash.c [hunk 2]
```diff
@@ -24,12 +25,45 @@ typedef struct hash_cfg_config_t {
...
+ * \param[in]  hash_b      Hash B to compare
+ *
+ * \return     On success returns a number between 0.0 and 1.0 (included), otherwise negative
+ */
+RZ_API double rz_hash_ssdeep_compare(RZ_NONNULL const char *hash_a, RZ_NONNULL const char *hash_b) {
+	rz_return_val_if_fail(hash_a && hash_b, -1.0);
+	return rz_ssdeep_compare(hash_a, hash_b);
+}
+
+RZ_API ut32 rz_hash_xxhash(RZ_NONNULL const ut8 *input, size_t size) {
```
Comment (XVilka) [inline librz/hash/hash.c:78]: Doxygen

File: librz/hash/hash.c [hunk 3]
```diff
@@ -24,12 +25,45 @@ typedef struct hash_cfg_config_t {
...
 	rz_return_val_if_fail(input, 0);
 	return XXH32(input, size, 0);
 }
 
-RZ_API double rz_hash_entropy(RZ_NONNULL RzHash *rh, RZ_NONNULL const ut8 *data, ut64 len) {
+RZ_API double rz_hash_entropy(RZ_NONNULL const ut8 *data, ut64 len) {
```
Comment (XVilka) [inline librz/hash/hash.c:91]: Same

File: librz/hash/hash.c [hunk 4]
```diff
@@ -479,9 +513,11 @@ RZ_API RZ_OWN char *rz_hash_cfg_get_result_string(RZ_NONNULL RzHashCfg *md, RZ_N
 	HashCfgConfig *mdc = (HashCfgConfig *)rz_list_iter_get_data(it);
 	rz_return_val_if_fail(mdc, NULL);
 
-	if (!strncmp(name, "entropy", 7)) {
+	if (!strncmp(name, "entropy", strlen("entropy"))) {
```
Comment (XVilka) [inline librz/hash/hash.c:549]: You could use `rz_str_startswith()` instead

File: librz/hash/hash.c [hunk 5]
```diff
@@ -552,10 +588,12 @@ RZ_API RZ_OWN char *rz_hash_cfg_calculate_small_block_string(RZ_NONNULL RzHash *
 		return NULL;
 	}
 
-	if (!strncmp(name, "entropy", 7)) {
+	if (!strncmp(name, "entropy", strlen("entropy"))) {
```
Comment (XVilka) [inline librz/hash/hash.c:624]: Same

File: librz/include/rz_hash.h [hunk 6]
```diff
@@ -74,9 +74,11 @@ RZ_API RZ_OWN ut8 *rz_hash_cfg_calculate_small_block(RZ_NONNULL RzHash *rh, RZ_N
 RZ_API RZ_OWN char *rz_hash_cfg_calculate_small_block_string(RZ_NONNULL RzHash *rh, RZ_NONNULL const char *name, RZ_NONNULL const ut8 *buffer, ut64 bsize, RZ_NULLABLE ut32 *size, bool invert);
 RZ_API RZ_OWN char *rz_hash_cfg_randomart(RZ_NONNULL const ut8 *buffer, ut32 length, ut64 address);
 
-RZ_API ut32 rz_hash_xxhash(RZ_NONNULL RzHash *rh, RZ_NONNULL const ut8 *input, size_t size);
-RZ_API double rz_hash_entropy(RZ_NONNULL RzHash *rh, RZ_NONNULL const ut8 *data, ut64 len);
-RZ_API double rz_hash_entropy_fraction(RZ_NONNULL RzHash *rh, RZ_NONNULL const ut8 *data, ut64 len);
+RZ_API double rz_hash_ssdeep_compare(RZ_NONNULL const char *hash1, RZ_NONNULL const char *hash2);
+RZ_API char *rz_hash_ssdeep(RZ_NONNULL const ut8 *input, size_t size);
```
Comment (XVilka) [inline librz/include/rz_hash.h]: Ownership

**Review by XVilka — CHANGES_REQUESTED** [review]
Looks good overall, you might want to update also a man page of `rz-hash` in `binrz/man/rz-hash.1`

File: librz/hash/hash.c (hunk 4 above)
Comment (wargio) [reply librz/hash/hash.c:549]: is not optimized.

**Comment by XVilka** [comment]
```

[XX] db/tools/rz_diff rz-diff similarity comparison (ssdeep) QUIET
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!rz-diff -i -d ssdeep '384:2rOP+jOA3orMecdFpCAFRa5FZiqboYOCE:CwqVCA/IFZiqboYOC' '384:lrOPfSTZjOA3orMecdFpu1FFl5bZiqboVOCX:zTZwqVu1fXbZiqboVOC'' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-0.696
+similarity: 0.696



[XX] db/tools/rz_diff rz-diff similarity comparison (ssdeep) json
RZ_NOPLUGINS=1 /usr/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '!rz-diff -i -d ssdeep '384:2rOP+jOA3orMecdFpCAFRa5FZiqboYOCE:CwqVCA/IFZiqboYOC' '384:lrOPfSTZjOA3orMecdFpu1FFl5bZiqboVOCX:zTZwqVu1fXbZiqboVOC'' =
-- stdout
--- expected
+++ actual
@@ -1,1 +1,1 @@
-{"similarity":0.695652}
+similarity: 0.696
```

**Review by XVilka — APPROVED** [review]

**Pull request merged**